design-docs/operationBased2.md

```diff
@@ -0,0 +1,269 @@
+<tr>
+<td><pre lang="graphql">
+query TestOperation {
+  animal {
+    ... on Cat {
+      species
+    }
+  }
+}
+</pre></td>
+<td><pre lang="kotlin">
+public data class Data(public val animal: IAnimal?)
+
+sealed interface IAnimal { val onCat: Animal.OnCat? }
+data class CatAnimal(val onCat: Animal.OnCat): IAnimal
+data class Animal(val onCat: Animal.OnCat?): IAnimal
+</pre></td>
+</tr>
+</table>
+
+
+
+## Example 6
+
+<table>
+<tr><th>GraphQL</th><th>Kotlin</th></tr>
+<tr>
+<td><pre lang="graphql">
+query TestOperation {
+  animal {
+    species
+    ... on WarmBlooded {
+      temperature
+    }
+    ... on Pet {
+      name
+    }
+  }
+}
+</pre></td>
+<td><pre lang="kotlin">
+public data class Data(public val animal: IAnimal?)
+
+// Abstraction Model
+sealed interface IAnimal { val onAnimal: Animal.IOnAnimal }
+// 3 TypeSet Models
+data class WarmBloodedAnimal(val onAnimal: Animal.IOnAnimal): IAnimal
+data class PetAnimal(val onAnimal: Animal.IOnAnimal): IAnimal
+data class WarmBloodedPetAnimal(val onAnimal: Animal.IOnAnimal): IAnimal
+// Fallback Model
+data class Animal(val onAnimal: IOnAnimal): IAnimal
+
+</pre></td>
+</tr>
+</table>
+
+
+## Example 7
+
+<table>
+<tr><th>GraphQL</th><th>Kotlin</th></tr>
+<tr>
+<td><pre lang="graphql">
+query TestOperation {
+  animal {
+    ... on Animal {
+      species
+      # Nested WarmBlooded
+      ... on WarmBlooded {
+        temperature
+      }
+      # Nested Pet
+      ... on Pet {
+        name
+      }
+    }
+  }
+}
+</pre></td>
+<td><pre lang="kotlin">
+
+// animal is not generated as polymorphic model because inline fragment ... on Animal is always assignable
+public data class Data(public val animal: Animal?)
+data class Animal(val onAnimal: IOnAnimal)
+
+// onAnimal field has two nested inline fragments -> generated as Polymorphic model
+public sealed interface IOnAnimal { 
+  val species: String,
+  val onWarmBlooded: OnAnimal.OnWarmBlooded?,
+   val onPet: OnAnimal.OnPet?
+}
+data class WarmBloodedOnAnimal(
+  val species: String, 
+  val onWarmBlooded: OnAnimal.OnWarmBlooded,
+  val onPet: OnAnimal.OnPet?
+): IOnAnimal
+data class PetOnAnimal(
+  val species: String,
+  val onWarmBlooded: OnAnimal.OnWarmBlooded?,
+  val onPet: OnAnimal.OnPet,
+): IOnAnimal
+data class WarmBloodedPetOnAnimal(
+  val species: String,
+  val onWarmBlooded: OnAnimal.OnWarmBlooded,
+  val onPet: OnAnimal.OnPet,
+): IOnAnimal
+data class OnAnimal(
+  val species: String,
+  val onWarmBlooded: OnWarmBlooded?,
+  val onPet: OnPet?,
+): IOnAnimal
+</pre></td>
+</tr>
+</table>
+
+## Example 8
+
+<table>
+<tr><th>GraphQL</th><th>Kotlin</th></tr>
+<tr>
+<td><pre lang="graphql">
+query TestOperation {
+  animal {
+    ... AnimalFragment
+  }
+}
+fragment AnimalFragment on Animal {
+  species
+  ... on WarmBlooded {
+    temperature
+  }
+  ... on Pet {
+    name
+  }
+}
+</pre></td>
+<td><pre lang="kotlin">
+
+// animal is not generated as polymorphic model because there is no inline fragment
```

Comment:
Can you elaborate why you only want this new codegen for inline fragments and not fragment spreads? With something like this:

```graphql
{
  animal {
    ...catFragment
    ...lionFragment
  }
}
```

it feels like the benefits would be the same as with 

```graphql
{
  animal {
    ... on Cat { ... }
    ... on Lion { ... }
  }
}
```

Any reason to only do the transformation for inline fragments?